### ISSUE TITLE
RFC: `#[serde(other)]` design

### ISSUE THREAD

**Issue by gtsiam** [serde-rs/serde#2126]
I wanted to take a crack at fully solving #912, but there are a couple of open questions I run into that I'd like to see comments on before I fully implement. Given the wide use of serde, the stability it has to provide ~~and the fact that I wrote most of this while trying to figure the problem out anyway~~, I think this is somewhat required to make a potentially controversial change (plus I don't want to implement it only to have it rejected). So with that out of the way:

## Terminology

- Variant: Rust enum variant
- Variant value: Data associated with a variant (unit, tuple or struct)
- "Other" variant: (note the quotation marks) The enum variant tagged with `#[serde(other)]`
- Tag: Name of the variant to parse
- Content: Data associated with a variant to parse

## Summary

Acceptable "Other" variants:

- Unit variant (This is unchanged)
- Newtype: (tag)
- 2-tuple: (tag, content)

The tag and content of the enum will be deserialized into the appropriate fields if they are present.

### Unit variant

The unit variant case should remain unchanged for compatibility reasons, which means that it will not store the tag and that it will expect a unit as its content.

### Tag

When a tag field is specified, it will be used to store the tag on deserialization and to set the tag on serialization. This would ensure that no information is lost when propagating unknown enums. If no tag information is stored, the variant name will be used for serialization same as any other variant.

### Content

When a content field is specified, it will be used for deserialization, otherwise a unit will be required. The content can be ignored explicitly with `serde::de::IgnoredAny`. This is not done by default for consistency with the unit variant case.

### Representing the tag

Serde supports a few different forms of tags (Mainly integers, strings and bytes). The requirement here is that the tag type can be (de)serialized from/to the tag used in the stored data (eg. String and &str should work for json, since serde_json uses strings to signify tags)

### Relation to `field_identifier`

A limited form of the aforementioned newtype `#[serde(other)]` variant is already supported by making the last variant of a `#[serde(field_identifier)]` a newtype over a tag. This approach should be consistent with that.

### History

First of all, #1382 has already implemented an (almost) complete version of `#[serde(other)]` which works by parsing the "other" variant same as any other variant. However that behavior is not currently usable on master due to the `other` attribute being limited to unit variants.

The reason for said limitations was due to unresolved questions around how to recover the original tag and some unintuitive behavior in a few common use cases. First, since the value of the "other" variant corresponds directly to the content of the un-parsed enum, there is nowhere to store the tag. @roblabla proposed to add a `#[serde(other_tag)]` option on one of the fields in the "other"
variant's value, which is what I was going to implement initially, however a few issues arise:

### Unrepresentable variants

In serde, variants with no values (unit variants), tuple, newtype and struct variants are handled differently. That means that:

```rs
// Parses:
// - [x] "CompilerMessage"
// - [x] { "CompilerMessage": null }
// - [ ] { "CompilerMessage": [] }
enum Message {
    #[serde(other)]
    Unknown
}
```

and

```rs
// Parses:
// - [ ] "CompilerMessage"
// - [ ] { "CompilerMessage": null }
// - [x] { "CompilerMessage": [] }
enum Message {
    #[serde(other)]
    Unknown()
}
```

are not the same. So that raises the question: After removing the `#[serde(other_tag)]` from the content, what form should the variant try to parse from? Assuming we do the obvious:

```rs
enum Message {
    #[serde(other)]
    Unknown(#[serde(other_tag)] String) // -> Tries `Unknown()`
}
```

Then that means that you can't pase the simple string case

### Unintuitive behavior

Derived from @dtolney's [comment](https://github.com/serde-rs/serde/pull/1382#issuecomment-420016369) on the original pr:

#### Non-obvious fail case

```rs
#[derive(Deserialize, PartialEq, Debug)]
enum E {
    A,
    #[serde(other)]
    Other(String), // This would require the #[serde(other_tag)]
                   // attribute, but it is not obvious how or why
}

fn main() {
    assert_eq!(
        serde_json::from_str::<E>(r#" "foo" "#).unwrap(),
        E::Other("foo".to_owned()),
    );
}
```

## Open questions

- This approach would require the tag and content to be part of the public api. Though the exact details could be hidden through newtypes over the tag and content, adding or removing a tag or content field could be a breaking change. I think this is acceptable, but this may be a point to discuss.
- In a past revision, I intended to implement one more "other" variant form: { tag, content }, but I removed it due to complications with how it would interact with struct fields. Maybe for struct variants, it'd be a good idea to support some form of `#[serde(other_tag)]`?

## Final note

My reservations revolve mostly around the question of how intuitive this all is, so I'd appreciate any feedback.

Edit History
---
- Removed the `{ tag, content }` variant form due to complications with field attributes. Leaving it up to future a future pr.

**Comment by gtsiam** [comment]
So I got deserialization somewhat working a few days ago, but serialization is a problem because the `Serializer::serialize_*_variant()` methods expect the variant name & index. Before I saw this, I expected I would be able to simply serialize the tag but turns out that is not the case. So I'm not really sure how to proceed.

- Leave the tag be `Other` (aka the actual variant name) and...
    - Serialize the captured content? But I can see that being very surprising behavior.
    - Discard the captured content and serialize as a unit variant? I'm not sure if that's acceptable
- Require "other" variants that capture a tag to be tagged with `#[serde(skip_serializing)]`. But that does not mirror current behavior at all.
- Refuse to serialize "other" variants altogether with captured data all-together (though that should be possible to work-around with `#[serde(serialize_with = "...")]`

I'm leaning towards a combination of the last two options because they seem to be the most forward-compatible and permissive ones, but this isn't what I had in mind when I started this... It's either that or discarding everything and serializing as we do now. The former would be a lot of pain for users and the latter would be more ergonomic though maybe surprising.

To be honest, I'm really not sure what the right choice is here. Neither is really - the right thing would be to adjust the `Serializer` trait. But after some browsing of the issue tracker, it looks like serde is now essentially ~~dead~~ stable, so this doesn't look likely.

There are also complications with the fact that (turns out) new type variants can't emulate all types of variants - `Variant` and `Variant(())` are two very different things, though I can accept that as an unavoidable limitation.
